cli/azd/pkg/commands/pipeline/azdo_provider.go

```diff
@@ -7,65 +7,528 @@ import (
 	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
 	"github.com/azure/azure-dev/cli/azd/pkg/infra/provisioning"
 	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/git"
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/build"
+	azdoGit "github.com/microsoft/azure-devops-go-api/azuredevops/git"
 )
 
-// AzdoHubScmProvider implements ScmProvider using Azure DevOps as the provider
+// AzdoScmProvider implements ScmProvider using Azure DevOps as the provider
 // for source control manager.
-type AzdoHubScmProvider struct {
+type AzdoScmProvider struct {
+	repoDetails    *AzdoRepositoryDetails
+	Env            *environment.Environment
+	AzdContext     *azdcontext.AzdContext
+	azdoConnection *azuredevops.Connection
+}
+
+// AzdoRepositoryDetails provides extra state needed for the AzDo provider.
+// this is stored as the details property in repoDetails
+type AzdoRepositoryDetails struct {
+	projectName     string
+	projectId       string
+	repoId          string
+	orgName         string
+	repoName        string
+	repoWebUrl      string
+	remoteUrl       string
+	sshUrl          string
+	buildDefinition *build.BuildDefinition
 }
 
 // ***  subareaProvider implementation ******
 
 // requiredTools return the list of external tools required by
 // Azure DevOps provider during its execution.
-func (p *AzdoHubScmProvider) requiredTools() []tools.ExternalTool {
+func (p *AzdoScmProvider) requiredTools() []tools.ExternalTool {
 	return []tools.ExternalTool{}
 }
 
 // preConfigureCheck check the current state of external tools and any
 // other dependency to be as expected for execution.
-func (p *AzdoHubScmProvider) preConfigureCheck(ctx context.Context, console input.Console) error {
-	return errors.New("not implemented")
+func (p *AzdoScmProvider) preConfigureCheck(ctx context.Context, console input.Console) error {
+	_, err := azdo.EnsurePatExists(ctx, p.Env, console)
+	if err != nil {
+		return err
+	}
+
+	_, err = azdo.EnsureOrgNameExists(ctx, p.Env, console)
+	return err
+}
+
+// helper function to save configuration values to .env file
+func (p *AzdoScmProvider) saveEnvironmentConfig(key string, value string) error {
+	p.Env.Values[key] = value
+	err := p.Env.Save()
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 // name returns the name of the provider
-func (p *AzdoHubScmProvider) name() string {
+func (p *AzdoScmProvider) name() string {
 	return "Azure DevOps"
 }
 
 // ***  scmProvider implementation ******
 
+// stores repo details in state for use in other functions. Also saves AzDo project details to .env
+func (p *AzdoScmProvider) StoreRepoDetails(ctx context.Context, repo *azdoGit.GitRepository) error {
+	repoDetails := p.getRepoDetails()
```

Comment:
For what it's worth, this pattern feels a little strange to me.  It seems like we have a bunch of places were we call getRepoDetails() and then perturb the returned object.  I wonder if there is some sort of structuring we can do long term such that we don't have this initialization split across all the files and most code can just access p.repoDetails directly.

Reply:
It took me a moment to work out that the value returned by repoDetails is a pointer to object state, so the writes on L90-94 were initializing shared state of the object.

Reply:
https://github.com/Azure/azure-dev/issues/756